### PR TITLE
Core IPC messaging rewrite.

### PR DESCRIPTION
**Pull request by kinetiknz**
Replaces tokio 0.1 with mio 0.7-based system.  Full description in [old draft PR](https://github.com/mozilla/audioipc-2/pull/123).

Part 1 includes the new IPC core, intended to replace the existing tokio 0.1 IPC core.  Part 2 integrates the new IPC core into the rest of the server and client code.  Part 3 removes the old tokio 0.1 core (including resolving mio 0.6 vs 0.7 dependency duplication).

These changes will be initially merged into the https://github.com/mozilla/audioipc-2/tree/macos branch, which can then be integrated with Gecko for the macOS build only.  Once any remaining issues have been resolved, the macos branch will be merged to master and Linux and Windows will switch to the new IPC core.

### CODE REVIEW

**Review by ChunMinChang — COMMENTED** [review]
I haven't read the whole code yet. I'll finish it next week. Just leaving some thoughts about the documents for now and some comments before I took the PTO last week.

File: audioipc/src/rpccore.rs
Comment (ChunMinChang) [inline audioipc/src/rpccore.rs:110]: It would be great if we can have some comments stating the relationship between `ClientHandler` and `ServerHandler`:
```
ClientHandler       ServerHandler
    produce ----------> consume 
    consume <---------- produce
```

File: audioipc/src/ipccore.rs
Comment (ChunMinChang) [inline audioipc/src/ipccore.rs:84]: nit: I find using `Ok(self.add_connection(server_pipe, driver)?)` and adding a `trace!(..)` printing the token here helpful for me to understand the code better. But I don't have a strong opinion.

**Review by ChunMinChang — COMMENTED** [review]
The change here is huge. Here are some initial thoughts after reading the code for now. I'll need to look deeper.

File: server/src/server.rs
Comment (ChunMinChang) [inline server/src/server.rs:323]: This is unrelated to the changes here, but could we rename this to `DeviceCollectionChangeCallback` or `DeviceCallback`, or something similar? `callback` appears in lots of the classes or variables in this crate, and each of them has a different meaning. I find it hard to read know what this is without jumping into the definitions. This bothers me when reading the code here.

File: client/src/context.rs
Comment (ChunMinChang) [inline client/src/context.rs:171]: I am not sure why this is named as `server_stream`. `ServerStream` is the cubeb stream on the ipc server-side/parent process. This should be the handle of `client_pipe` created in ipc server-side. Could we name this `server_connection`?

**Review by ChunMinChang — CHANGES_REQUESTED** [review]
The readability is a lot better now, awesome! I enjoy reading the changes proposed here. I'm just holding this off until the questions here are answered. Some of the comments are unrelated to the changes made here but I think it's worth addressing those issues while we are trying to make the code easier to be maintained.

File: ARCHITECTURE.md
Comment (kinetiknz) [reply ARCHITECTURE.md:31]: Good idea.  I've removed the TODOs since they're complete (and it's better to track new ones in issues or as a TODO in the code for small items), and moved the contents of  ARCHITECTURE to README (with some rewording).

File: ARCHITECTURE.md
Comment (kinetiknz) [reply ARCHITECTURE.md:25]: Rewritten per suggestions, hopefully it's somewhat clearer.  Happy to further add or clarify things in this PR as needed, but I'd also like continue expanding this document in a later PR along with reworking naming conventions of various parts of the code to reduce the confusion between the multiple uses of server, client, and other overloaded terms.

File: audioipc/src/ipccore.rs
Comment (kinetiknz) [reply audioipc/src/ipccore.rs:128]: Renamed for clarity.

File: audioipc/src/ipccore.rs
Comment (kinetiknz) [reply audioipc/src/ipccore.rs:298]: Done.

File: audioipc/src/ipccore.rs
Comment (kinetiknz) [reply audioipc/src/ipccore.rs:602]: Removed.

File: audioipc/src/ipccore.rs
Comment (kinetiknz) [reply audioipc/src/ipccore.rs:621]: Removed.

File: audioipc/src/ipccore.rs
Comment (kinetiknz) [reply audioipc/src/ipccore.rs:542]: All existing callers supply `F1` and `F2`.  Making them optional clutters the call site since you need to explicitly specify the type of `F1` and `F2` in the `None` case - that could be solved with a builder pattern, but since they're currently always specified it doesn't seem worth the added complexity.

File: audioipc/src/ipccore.rs
Comment (kinetiknz) [reply audioipc/src/ipccore.rs:653]: Good idea, done.

File: audioipc/src/rpccore.rs
Comment (kinetiknz) [reply audioipc/src/rpccore.rs:109]: That's tidier, thanks.

File: audioipc/src/rpccore.rs
Comment (kinetiknz) [reply audioipc/src/rpccore.rs:173]: Done.

File: audioipc/src/rpccore.rs
Comment (kinetiknz) [reply audioipc/src/rpccore.rs:45]: Renamed these to `Response` and `Request` everywhere.

File: audioipc/src/rpccore.rs
Comment (kinetiknz) [reply audioipc/src/rpccore.rs:127]: Agree it's confusing.  I've tried renaming the types to ClientMessage and ServerMessage where appropriate; I think it's a bit clearer now.

File: audioipc/src/rpccore.rs
Comment (kinetiknz) [reply audioipc/src/rpccore.rs:44]: Done.

File: audioipc/src/ipccore.rs
Comment (kinetiknz) [reply audioipc/src/ipccore.rs:394]: Yeah, processing inbound messages will result in outbound messages being produced, which we then want to send on the connection.

File: audioipc/src/rpccore.rs
Comment (kinetiknz) [reply audioipc/src/rpccore.rs:148]: Done.

File: audioipc/src/rpccore.rs
Comment (kinetiknz) [reply audioipc/src/rpccore.rs:149]: Let's consider adding this later if it'll be useful for debugging.

File: audioipc/src/ipccore.rs
Comment (kinetiknz) [reply audioipc/src/ipccore.rs:170]: Good idea, added.

File: audioipc/src/ipccore.rs
Comment (kinetiknz) [reply audioipc/src/ipccore.rs:290]: `Connection` doesn't have access to the `poll` on `EventLoop`, so not directly.  It'd be possible to use `Registry::try_clone` to store a registry instance per connection, but I'm not sure if that's an improvement over passing in a reference.

File: audioipc/src/ipccore.rs
Comment (kinetiknz) [reply audioipc/src/ipccore.rs:291]: Thanks for the suggestion!  I've refactored `handle_event` into smaller functions, hopefully it's more readable now.

File: audioipc/src/ipccore.rs
Comment (kinetiknz) [reply audioipc/src/ipccore.rs:84]: Done.

File: server/src/server.rs
Comment (kinetiknz) [reply server/src/server.rs:555]: Sorry, that was a badly worded comment but you've understood it correctly (`NamedPipe` lacks an `into_raw_handle`).  I've tried to improve the comment.

Note that the old code also passed the "wrong" pipe here for similar reasons, so the behaviour should be unchanged.

File: audioipc/src/sys/mod.rs
Comment (kinetiknz) [reply audioipc/src/sys/mod.rs:36]: Oops, fixed.

File: audioipc/src/ipccore.rs
Comment (kinetiknz) [reply audioipc/src/ipccore.rs:303]: `IPC_CLIENT_BUFFER_SIZE` is an arbitrary starting buffer size, it's not expected to be large enough for all possible messages.  When `decode` is called, it will always ensure there's sufficient space available to read the next message length prefix (at message start) or next entire message (once message length is known from the length prefix).  The EventLoop will continue reading and attempting to decode as long as there's data available on the connection, causing `decode` to adjust the buffer size as necessary.

It'd be better if the buffer size management was clearer, I'll think about how it might be improved.

File: audioipc/src/sys/mod.rs
Comment (kinetiknz) [reply audioipc/src/sys/mod.rs:38]: Thanks, removed.

File: audioipc/src/sys/mod.rs
Comment (kinetiknz) [reply audioipc/src/sys/mod.rs:44]: Thanks, removed.

File: audioipc/src/sys/unix/mod.rs
Comment (kinetiknz) [reply audioipc/src/sys/unix/mod.rs:45]: Moved msg.rs (and cmsg.rs) into audioipc/src/sys/unix.

File: audioipc/src/sys/unix/mod.rs
Comment (kinetiknz) [reply audioipc/src/sys/unix/mod.rs:66]: Moved msg.rs (and cmsg.rs) into audioipc/src/sys/unix.

File: audioipc/src/sys/unix/mod.rs
Comment (kinetiknz) [reply audioipc/src/sys/unix/mod.rs:72]: `b.is_empty` is true for any messages without an associated handle being sent.  Since this layer doesn't know anything about messages, it needs to handle both cases.  When `b.is_empty` is true, `cmsg::iterator` returns `None`.

File: audioipc/src/ipccore.rs
Comment (kinetiknz) [reply audioipc/src/ipccore.rs:465]: Related to the answer above for `send_msg` - `b.is_empty` is true when the received message has no associated handle.

File: server/src/server.rs
Comment (kinetiknz) [reply server/src/server.rs:323]: That's clearer, I've renamed it.

File: client/src/context.rs
Comment (kinetiknz) [reply client/src/context.rs:171]: Done.

File: audioipc/src/rpccore.rs
Comment (kinetiknz) [reply audioipc/src/rpccore.rs:119]: I've renamed this to `messages`, does that seem more suitable?

File: server/src/server.rs
Comment (kinetiknz) [reply server/src/server.rs]: Agreed.  I can't recall why this was using the callback thread in the old code.  I've moved it to the RPC thread now.

**Comment by kinetiknz** [comment]
Thanks for the great feedback!  Really appreciate the time and thought you've put in to this.  I've addressed most of your comments now (forgot to re-request review when I pushed the bulk of the changes addressing your comments on Monday. 🤦‍♂️).  I'll keep working on the few that are still outstanding and put them up soon.

I've also included a series of small fixes and improvements as individual changes.  I'll squash those into the main commits before merging.

File: ARCHITECTURE.md
Comment (kinetiknz) [reply ARCHITECTURE.md]: https://github.com/mozilla/audioipc-2/issues/133

File: audioipc/src/codec.rs
Comment (ChunMinChang) [inline audioipc/src/codec.rs:163]: Is the performance bottleneck here? Is it related to memory allocations?

**Review by ChunMinChang — APPROVED** [review]
Great improvements! Thanks a lot!

File: ipctest/src/main.rs
Comment (kinetiknz) [reply ipctest/src/main.rs:185]: It's reading a whole line, so it won't unblock until it sees a newline.  It's true that it doesn't check the input, so any keys pressed before enter are ignored.

File: audioipc/src/codec.rs
Comment (kinetiknz) [reply audioipc/src/codec.rs:163]: There are speed ups in both `encode` and `decode`.  I noticed these while upgrading to bytes 1.1 (not included, needs more work) where I accidentally made serialization slow.  I didn't profile before/after (speed up numbers are from the included criterion benchmark), so I don't have a precise explanation for where the improvements are.  In `decode` it seems to be mainly from avoiding the split & freeze.  Same with `encode` (using the extra buffer to calculate the length prefix after encoding), plus switching from passing a `Writer` in favour of the extra buffer into `serialize_into`.

**Pull request merged**
